lib/pure/collections/smart_ptr.nim

```diff
@@ -0,0 +1,152 @@
+    p.val.allocator.dealloc(p.val.allocator, p.val, sizeof(p.val[]))
+    p.val = nil
+
+proc `=`*[T](dest: var UniquePtr[T], src: UniquePtr[T]) {.error.}
+
+proc `=sink`*[T](dest: var UniquePtr[T], src: UniquePtr[T]) {.inline.} =
+  if dest.val != src.val:
+    if dest.val != nil:
+      dest.val = src.val
+
+proc newUniquePtr*[T](val: sink T): UniquePtr[T] =
+  let a = getSharedAllocator()
+  result.val = cast[type(result.val)](a.alloc(a, sizeof(result.val[])))
+  if AllocatorFlag.ZerosMem notin a.flags:
+    reset(result.val[])
+  result.val.value = val
+  result.val.allocator = a
+
+converter convertUniquePtrToObj*[T](p: UniquePtr[T]): var T {.inline.} =
+  p.val.value
+
+proc isNil*[T](p: UniquePtr[T]): bool {.inline.} =
+  p.val == nil
+
+proc get*[T](p: UniquePtr[T]): var T {.inline.} =
+  when compileOption("boundChecks"):
+    if p.val == nil:
+      raise newException(ValueError, "deferencing nil unique pointer")
+  p.val.value
+
+proc `$`*[T](p: UniquePtr[T]): string {.inline.} =
+  if p.val == nil: "UniquePtr[" & $T & "](nil)"
+  else: "UniquePtr[" & $T & "](" & $p.val.value & ")"
+
+#------------------------------------------------------------------------------
+
+type
+  SharedPtr*[T] = object
+    ## shared ownership reference counting pointer
+    val: ptr tuple[value: T, atomicCounter: int, allocator: Allocator]
+
+proc `=destroy`*[T](p: var SharedPtr[T]) =
+  mixin `=destroy`
+  if p.val != nil:
+    let c = atomicDec(p.val[].atomicCounter)
+    if c == 0:
+      `=destroy`(p.val[])
+      p.val.allocator.dealloc(p.val.allocator, p.val, sizeof(p.val[]))
+    p.val = nil
+
+proc `=sink`*[T](dest: var SharedPtr[T], src: SharedPtr[T]) {.inline.} =
+  if dest.val != src.val:
+    if dest.val != nil:
+      `=destroy`(dest)
+    dest.val = src.val
+
+proc `=`*[T](dest: var SharedPtr[T], src: SharedPtr[T]) =
+  if dest.val != src.val:
+    if dest.val != nil:
+      `=destroy`(dest)
+    if src.val != nil:
+      discard atomicInc(src.val[].atomicCounter)
+      dest.val = src.val
+
+proc newSharedPtr*[T](val: sink T): SharedPtr[T] =
+  let a = getSharedAllocator()
+  result.val = cast[type(result.val)](a.alloc(a, sizeof(result.val[])))
+  if AllocatorFlag.ZerosMem notin a.flags:
+    reset(result.val[])  
+  result.val.value = val
+  result.val.atomicCounter = 1
+  result.val.allocator = a
+
+converter convertSharedPtrToObj*[T](p: SharedPtr[T]): var T {.inline.} =
+  p.val.value
+
+proc isNil*[T](p: SharedPtr[T]): bool {.inline.} =
+  p.val == nil
+
+proc get*[T](p: SharedPtr[T]): var T {.inline.} =
+  when compileOption("boundChecks"):
+    if p.val == nil:
+      raise newException(ValueError, "deferencing nil shared pointer")
+  p.val.value
+
+proc `$`*[T](p: SharedPtr[T]): string {.inline.} =
+  if p.val == nil: "SharedPtr[" & $T & "](nil)"
+  else: "SharedPtr[" & $T & "](" & $p.val.value & ")"
+
+#------------------------------------------------------------------------------
+
+type
+  ConstPtr*[T] = distinct SharedPtr[T]
```

Comment:
Doesn't `distinct` here mean that the user can easily erase the immutability by converting to `SharedPtr`?

Reply:
Sure but that's ok with me.